dag_in_context/src/interval_analysis.egg

```diff
@@ -1,87 +1,233 @@
+(rewrite (bound-min (IntB x) (IntB y))
+         (IntB (min x y))
          :ruleset interval-analysis)
-(rewrite (interval-intersect (BoolI la ha) (BoolI lb hb))
-         (BoolI (or la lb) (and ha hb))
+(rewrite (bound-max (BoolB x) (BoolB y))
+         (BoolB (or x y))
          :ruleset interval-analysis)
-(rewrite (interval-union (BoolI la ha) (BoolI lb hb))
-         (BoolI (and la lb) (or ha hb))
+(rewrite (bound-min (BoolB x) (BoolB y))
+         (BoolB (and x y))
          :ruleset interval-analysis)
 
-; Interval Table
-(function ival (Expr) Interval
-  :merge (interval-intersect old new))
-
 ; =================================
 ; Constants
 ; =================================
 (rule ((= lhs (Const (Int x) ty)))
-      ((set (ival lhs) (IntI x x)))
+      (
+        (set (lo-bound lhs) (IntB x))
+        (set (hi-bound lhs) (IntB x))
+      )
       :ruleset interval-analysis)
 
 (rule ((= lhs (Const (Bool x) ty)))
-      ((set (ival lhs) (BoolI x x)))
+      (
+        (set (lo-bound lhs) (BoolB x))
+        (set (hi-bound lhs) (BoolB x))
+      )
       :ruleset interval-analysis)
 
 ; =================================
 ; Constant Folding
 ; =================================
 (rule (
-       (= (IntI x x) (ival expr))
+       (= (IntB x) (lo-bound expr))
+       (= (IntB x) (hi-bound expr))
        (HasArgType expr ty)
       )
       ((union expr (Const (Int x) ty)))
       :ruleset interval-analysis)
 
 (rule (
-       (= (BoolI x x) (ival expr))
+       (= (BoolB x) (lo-bound expr))
+       (= (BoolB x) (hi-bound expr))
        (HasArgType expr ty)
       )
       ((union expr (Const (Bool x) ty)))
       :ruleset interval-analysis)
 
+; lower bound being true means the bool must be true
+(rule (
+       (= (BoolB true) (lo-bound expr))
+       (HasArgType expr ty)
+      )
+      ((union expr (Const (Bool true) ty)))
+      :ruleset interval-analysis)
+
+; upper bound being false means the bool must be false
+(rule (
+       (= (BoolB false) (hi-bound expr))
+       (HasArgType expr ty)
+      )
+      ((union expr (Const (Bool false) ty)))
+      :ruleset interval-analysis)
+
 ; =================================
 ; Arithmetic
 ; =================================
 ; + a b interval is (+ la lb) (+ ha hb)
 (rule (
-       (= lhs (Bop (Add) x y))
-       (= (ival x) (IntI la ha))
-       (= (ival y) (IntI lb hb))
+       (= lhs (Bop (Add) a b))
+       (= (IntB la) (lo-bound a))
+       (= (IntB lb) (lo-bound b))
+      )
+      (
+       (set (lo-bound lhs) (IntB (+ la lb)))
+      )
+      :ruleset interval-analysis)
+(rule (
+       (= lhs (Bop (Add) a b))
+       (= (IntB ha) (hi-bound a))
+       (= (IntB hb) (hi-bound b))
+      )
+      (
+       (set (hi-bound lhs) (IntB (+ ha hb)))
       )
-      ((set (ival lhs) (IntI (+ la hb) (+ ha hb))))
+      :ruleset interval-analysis)
+
+; mul commutes
+(rewrite (Bop (Mul) x y) (Bop (Mul) y x) :ruleset interval-analysis)
+
+; negative number * negative number is positive
+(rule (
```

Comment:
Seems like you are missing some Mul rules- perhaps in the next PR?
These handle pretty specific cases. The intervals may also be mixed sign (not strictly positive or negative)